pkg/build/controller/build/defaults/defaults.go

```diff
@@ -110,44 +111,59 @@ func (b BuildDefaults) applyPodDefaults(pod *v1.Pod, isCustomBuild bool) {
 	for _, c := range allContainers {
 		// All env vars are allowed to be set in a custom build pod, the user already has
 		// total control over the env+logic in a custom build pod anyway.
+		externalEnv := make([]corev1.EnvVar, len(b.Config.Env))
+		for i, v := range b.Config.Env {
+			externalEnv[i] = corev1.EnvVar{}
+			if err := legacyscheme.Scheme.Convert(&v, &externalEnv[i], nil); err != nil {
+				panic(err)
+			}
+		}
 		if isCustomBuild {
-			util.MergeEnvWithoutDuplicates(util.CopyApiEnvVarToV1EnvVar(b.Config.Env), &c.Env, false, []string{})
+			util.MergeEnvWithoutDuplicates(externalEnv, &c.Env, false, []string{})
 		} else {
-			util.MergeTrustedEnvWithoutDuplicates(util.CopyApiEnvVarToV1EnvVar(b.Config.Env), &c.Env, false)
+			util.MergeTrustedEnvWithoutDuplicates(externalEnv, &c.Env, false)
 		}
 
 		if c.Resources.Limits == nil {
-			c.Resources.Limits = v1.ResourceList{}
+			c.Resources.Limits = corev1.ResourceList{}
 		}
 		for name, value := range defaultResources.Limits {
-			if _, ok := c.Resources.Limits[v1.ResourceName(name)]; !ok {
+			if _, ok := c.Resources.Limits[corev1.ResourceName(name)]; !ok {
 				glog.V(5).Infof("Setting default resource limit %s for pod %s/%s to %v", name, pod.Namespace, pod.Name, value)
-				c.Resources.Limits[v1.ResourceName(name)] = value
+				c.Resources.Limits[corev1.ResourceName(name)] = value
 			}
 		}
 		if c.Resources.Requests == nil {
-			c.Resources.Requests = v1.ResourceList{}
+			c.Resources.Requests = corev1.ResourceList{}
 		}
 		for name, value := range defaultResources.Requests {
-			if _, ok := c.Resources.Requests[v1.ResourceName(name)]; !ok {
+			if _, ok := c.Resources.Requests[corev1.ResourceName(name)]; !ok {
 				glog.V(5).Infof("Setting default resource request %s for pod %s/%s to %v", name, pod.Namespace, pod.Name, value)
-				c.Resources.Requests[v1.ResourceName(name)] = value
+				c.Resources.Requests[corev1.ResourceName(name)] = value
 			}
 		}
 	}
 }
 
-func (b BuildDefaults) applyBuildDefaults(build *buildapi.Build) {
+func (b BuildDefaults) applyBuildDefaults(build *buildv1.Build) {
 	// Apply default env
 	for _, envVar := range b.Config.Env {
 		glog.V(5).Infof("Adding default environment variable %s=%s to build %s/%s", envVar.Name, envVar.Value, build.Namespace, build.Name)
-		addDefaultEnvVar(build, envVar)
+		externalEnv := corev1.EnvVar{}
+		if err := legacyscheme.Scheme.Convert(&envVar, &externalEnv, nil); err != nil {
+			panic(err)
```

Comment:
panic?

Reply:
there is no error reporting from that function up and failing to convert here means you are silently loosing that var.

Reply:
still, isn't that better than panicking the entire controller process if something goes wrong here?

Reply:
(you could log it so we at least have evidence in the controller logs that it happened)

Reply:
@bparees if you really need some vars in builds and won't get them but 1000 builds are triggered, do you want to investigate after the 1000 builds ran or you want to prevent those builds from being triggered?

Reply:
i'd rather investigate after than get a call at 2am because some malicious user found a way to force the controller process to panic by submitting builds w/ non-convertable env vars (or some other bug in our conversion logic).

alternatively you can add an error return on this function, after all the real entrypoint is `ApplyDefaults` which can bubble the error up, so why not just do that?